### PR TITLE
Add "GetLayoutAsync" command, "Hostname" getter, Optional HttpClient input

### PR DESCRIPTION
**Pull request by d8ahazard**
Add option to retrieve panel layout...as I use that one a lot.
Add getter for Hostname, so it can be read from the client at a later time if desired.
Add option to pass an existing HttpClient to LifxClient constructor. I did this because in my app, I have many devices which use a HttpClient, and found it is more efficient to just cache one and use it everywhere, versus creating a new one for each device.
I also removed the two-param constructor for LifxClient, and just created one with three values that have optional fallbacks - meaning nothing should change at all user-side in terms of creating a new client.

### CODE REVIEW

**Comment by d8ahazard** [comment]
Also, as the last commit indicates...add support for streaming. ;)

Honestly, I've been using your library for a while now, but apparently just copied a big chunk of the code into my app, and then used your methods to do the "extra" stuff. This is basically just me re-offloading that code to your library and adding the "extra" stuff I was doing, in a cleaner form that what I had implemented.

Everything has been tested with my Canvas tiles and works as desired.

**Comment by d8ahazard** [comment]
Hello, please let me know if you're still maintaining this package, or if I should create a new package and start maintaining it myself. I'd prefer to not have to do that, but I do need to include it in an app I'm releasing soon, and I would definitely like to have the code public.

Thanks!

**Comment by BullFrog13** [comment]
Hi @d8ahazard. Thanks for your contribution! Yes, I maintain the package in my free time.

I will take a look by the end of this week.

File: Nanoleaf.Client/Nanoleaf.Client/NanoleafHttpClient.cs [hunk 1]
```diff
@@ -7,15 +7,18 @@
 
 namespace Nanoleaf.Client
 {
-    internal class NanoleafHttpClient : HttpClient
```
Comment (BullFrog13) [inline Nanoleaf.Client/Nanoleaf.Client/NanoleafHttpClient.cs:10]: NanoleafClient is designed in a way that it disposes of the underlying NanoleafHttpClient, This change breaks the use of the Disposable pattern and introduces strong dependency in a form of HttpClient class.

I assume this change was made since you have some custom HTTP Client behaviour.

I suggest that you should extend the existing class. For example, introduce INanoleafHttpClient for NanoleafHttpClient that will be injected into NanoleafClient. Then, in your client library, you will be able to create your own HttpClient and extend it as you wish as well as NanoleafClient.

This would eliminate strong dependency and will allow you to change the behaviour of HttpClient.

Keep in mind that this library can not be a silver bullet for everything. While there are many cases and uses for it, the main goal is to provide the base that can be extended by clients if needed.

File: Nanoleaf.Client/Nanoleaf.Client/NanoleafClient.cs [hunk 2]
```diff
@@ -19,17 +20,23 @@ public class NanoleafClient : INanoleafClient
     {
         private bool _isDisposed;
         private readonly NanoleafHttpClient _nanoleafHttpClient;
+        /// <summary>
+        /// Allow us to retrieve the device hostname for later
+        /// </summary>
+        public string HostName { get; }
```
Comment (BullFrog13) [inline Nanoleaf.Client/Nanoleaf.Client/NanoleafClient.cs:26]: That's a good addition to the client, thanks

File: Nanoleaf.Client/Nanoleaf.Client/NanoleafClient.cs [hunk 3]
```diff
@@ -19,17 +20,23 @@ public class NanoleafClient : INanoleafClient
...
 
-        public NanoleafClient(string host)
+        /// <summary>
+        /// Create a new nanoleaf client
+        /// </summary>
+        /// <param name="host">Hostname or IP address of nanoleaf device</param>
+        /// <param name="userToken">(Optional) User token to use if authorized</param>
+        /// <param name="client">(Optional) Used to pass a shared HttpClient</param>
+        public NanoleafClient(string host, string userToken = "", HttpClient client = null)
```
Comment (BullFrog13) [inline Nanoleaf.Client/Nanoleaf.Client/NanoleafClient.cs:34]: Please, take a look at the comment below for NanoleafHttpClient class.

File: Nanoleaf.Client/Nanoleaf.Client/NanoleafStreamingClient.cs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+using System.Collections.Generic;
+using System.Drawing;
+using System.Globalization;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+
+namespace Nanoleaf.Client {
+	public class NanoleafStreamingClient : IDisposable {
```
Comment (BullFrog13) [inline Nanoleaf.Client/Nanoleaf.Client/NanoleafStreamingClient.cs]: I suggest using the same approach for this client as well. The pattern is the same.

File: Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+		/// </summary>
+		[JsonProperty] public PanelLayout[] PositionData { get; set; } = Array.Empty<PanelLayout>();
+
+	}
+
+	/// <summary>
+	/// Layout info on a specific panel
+	/// </summary>
+	[Serializable]
+	public class PanelLayout {
```
Comment (BullFrog13) [inline Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs]: We need to use separate files for different classes. Let's follow the existing code design.

File: Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+namespace Nanoleaf.Client.Models.Responses {
+	/// <summary>
+	/// Layout response returned from a layout request
+	/// </summary>
+	[Serializable]
+	public class Layout {
+		/// <summary>
+		/// Number of panels
+		/// </summary>
+		[JsonProperty] public int NumPanels { get; set; }
```
Comment (BullFrog13) [inline Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs]: Please use the existing code design. Attributes should be on a separate line from a property.

File: Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.ComponentModel;
+using Newtonsoft.Json;
+
+namespace Nanoleaf.Client.Models.Responses {
+	/// <summary>
+	/// Layout response returned from a layout request
+	/// </summary>
+	[Serializable]
+	public class Layout {
```
Comment (BullFrog13) [inline Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs]: Let's follow the existing code design. Curly braces come on the next line.

**Review by BullFrog13 — CHANGES_REQUESTED** [review]
Changes to design and refactoring are required.

File: Nanoleaf.Client/Nanoleaf.Client/NanoleafHttpClient.cs (hunk 1 above)
Comment (d8ahazard) [reply Nanoleaf.Client/Nanoleaf.Client/NanoleafHttpClient.cs:10]: So, the idea behind this one is simply to allow me to reuse an existing HTTPClient, if desired.

Functionally, nothing should be broken. If the user doesn't specify an existing HTTPClient, the library will still generate and use it's own, and still dispose of it when the Nanoleaf client is disposed. But, if we pass an existing client, the library knows not to dispose of it when the NanoleafClient is disposed.

Basically, for my purposes, this makes the most sense, as I then just have one HTTPClient my entire app references, which is shared by the various underlying libraries. I'm doing something similar for UDP Clients as well...

File: Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs (hunk 5 above)
Comment (d8ahazard) [reply Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs]: Fixed with 21f16ef, f39c588.

File: Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs (hunk 6 above)
Comment (d8ahazard) [reply Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs]: Fixed with 21f16ef, f39c588.

File: Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs (hunk 7 above)
Comment (d8ahazard) [reply Nanoleaf.Client/Nanoleaf.Client/Models/Responses/Layout.cs]: Fixed with 21f16ef.

File: Nanoleaf.Client/Nanoleaf.Client/NanoleafClient.cs (hunk 3 above)
Comment (d8ahazard) [reply Nanoleaf.Client/Nanoleaf.Client/NanoleafClient.cs:34]: See my comment for NanoleafHttpClient with remarks regarding this.

If you *really* want it removed, I can, but I'd humbly ask you to reconsider. It is an optional param, and shouldn't break any functionality...but it'd save me a bit of rewriting to have to undo.

**Review by BullFrog13 — APPROVED** [review]

File: Nanoleaf.Client/Nanoleaf.Client/NanoleafHttpClient.cs (hunk 1 above)
Comment (BullFrog13) [reply Nanoleaf.Client/Nanoleaf.Client/NanoleafHttpClient.cs:10]: I see your point. I agree with you. Lets proceed

**Comment by BullFrog13** [comment]
LGTM. This will go into V2.0 release
